src/utils.py

```diff
@@ -29,28 +37,71 @@ def format_token_short(token):
 
 
 def format_integer(number):
-  return str(number) # TODO: Format better the numbers
+  return '{:,d}'.format(number)
+
+def format_amount_in_weis(amount, decimals, rounding=ROUND_DOWN):
+  value = amount / Decimal(10 ** decimals)
+  return format_amount(value, decimals=decimals, rounding=rounding)
 
-def format_amount_in_weis(amount, decimals):
-  return str(amount / (10 ** decimals)) # TODO: Format better the weis
+def format_price(amount, decimals=10, rounding=ROUND_DOWN, currency=''):
+  price = format_amount(amount, decimals=decimals, rounding=rounding)
+  return price + ' ' + currency if currency else price
+
+def format_amount(amount, decimals=18, rounding=ROUND_DOWN):
+  quantize_value = Decimal(10) ** -Decimal(decimals)
+  rounded_value = Decimal(amount).quantize(quantize_value, rounding=rounding)
+
+  return f'{{:,.{decimals}f}}'.format(rounded_value).rstrip('0').rstrip('.')
 
 def format_date(date):
   return '' if date is None else date.strftime("%d/%m/%y")
 
-def format_date_time(date):
-  return '' if date is None else date.strftime("%d/%m/%y %H:%M:%S")
+def format_date_time(date, tooBigLabel='Never'):  
+    return tooBigLabel if date == datetime.max else date.strftime("%d/%m/%y %H:%M:%S")
+
+def format_batch_id_with_date(batchId, tooBigLabel='Never expires'):
+  if batchId:
+    return tooBigLabel if batchId >= MAX_BATCH_ID else (
+      format_integer(batchId) + 
+      f" ({ format_date_time(to_date_from_batch_id(batchId))})"
+    )
+  else:
+    return ''
 
 def to_date_from_epoch(epoch):
-  return epoch # TODO: Date from epoch
+  global MAX_EPOCH
+
+  if epoch < MAX_EPOCH:
+    return datetime.utcfromtimestamp(epoch) if epoch else None
+  else:
+    return datetime.max
 
 def to_etherscan_link(hash):
   return 'https://etherscan.io/tx/' + hash
 
-def toDateFromBatchId(batchId):
-  return batchId * BATCH_TIME_SECONDS # TODO: Dates in python
+def to_date_from_batch_id(batchId):
+  return to_date_from_epoch(batchId * BATCH_TIME_SECONDS)
 
 def calculate_price(numerator, denominator, decimals_numerator, decimals_denominator):
-  return numerator/denominator # TODO: Take decimals into account
+  numerator_dec = Decimal(numerator)
+  denominator_dec = Decimal(denominator)
+
+  if denominator_dec.is_zero():
+    return Decimal(1) if numerator_dec.is_zero() else Infinity
+  else:
+    precision_factor = Decimal(10) ** Decimal(abs(decimals_numerator - decimals_denominator))
+    if decimals_numerator > decimals_denominator:
+      return numerator_dec / denominator_dec / precision_factor
+    else:
+      return numerator_dec / (denominator_dec / precision_factor)
+
+  if decimals_numerator > decimals_denominator:
```

Comment:
Is this reachable?

Reply:
O myyyyy! I forgor to delete that last part! I'll address it here https://github.com/gnosis/dex-cli/issues/39

super good catch!